### PR TITLE
feat(rust, python): read ipc file from network

### PR DESCRIPTION
**Pull request by mickvangelderen**
Draft to discuss ongoing work from https://github.com/pola-rs/polars/issues/14839

### CODE REVIEW

File: crates/polars-io/src/ipc/object_store_reader.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    let file_bytes = with_concurrency_budget(1, || read_bytes(store, path)).await?;
+
+    let reader = IpcReader::new(std::io::Cursor::new(file_bytes));
+
+    let data = reader.finish()?;
+
+    PolarsResult::Ok(data)
+}
+
+async fn read_bytes<T: ObjectStore>(store: &T, path: &ObjectPath) -> PolarsResult<Bytes> {
```
Comment (ritchie46) [inline crates/polars-io/src/ipc/object_store_reader.rs:41]: This must eventually launch a thread that prefetches `ArrowChunk` and stores them on a queue.

The `AsyncIpcReader` must have a `batched` method that turns this reader into a `BatchedReader`.

The `BatchedReader` will have a `next_batches` that will fetch batches from that queues receiving end.

You can mostly copycat all fields/methods from parquet. E.g. we need to be able to apply a `limit`, project only a certain amount of columns, apply a filter etc.

**Comment by mickvangelderen** [comment]
@ritchie46 is it correct that we read the schema during the construction of the logical plan, and then only later during execution of the physical plan read the entire file?

If that is the case, I think I should provide a function that reads just the schema and one that reads the necessary data. These functions will take different options. For example, reading the schema wouldn't require projection names. Reading the data can take the projection indices computed from the schema + indices. There are probably other differences. Some options are probably be shared (some of the cloud-related options?).

File: crates/polars-lazy/src/physical_plan/executors/scan/ipc.rs [hunk 2]
```diff
@@ -12,14 +17,46 @@ pub struct IpcExec {
...
-        let file = std::fs::File::open(&self.path)?;
+        if is_cloud_url(&self.path) || env_force_async() {
+            #[cfg(not(feature = "cloud"))]
+            {
+                panic!("activate cloud feature")
+            }
+
+            #[cfg(feature = "cloud")]
+            {
+                // TODO: This will block the current thread until the data has been
```
Comment (ritchie46) [inline crates/polars-lazy/src/physical_plan/executors/scan/ipc.rs:28]: This is fine as the default engine cannot work on batches anyway. What our goal here will be is downloading parts of the file (e.g. `Recordbatches/ Chunks`) concurrently.

File: crates/polars-io/src/ipc/object_store_reader.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    let path = ObjectPath::parse(path).map_err(to_compute_err)?;
+
+    // TODO: Load only what is needed, rather than everything.
+    let file_bytes = with_concurrency_budget(1, || read_bytes(store, &path)).await?;
+    
+    read_ipc(std::io::Cursor::new(file_bytes), options)
+}
+
+async fn read_bytes<O: ObjectStore>(store: &O, path: &ObjectPath) -> PolarsResult<Bytes> {
+    // TODO: Is `to_compute_err` appropriate? It is used in the Parquet
```
Comment (ritchie46) [inline crates/polars-io/src/ipc/object_store_reader.rs:70]: Compute error is fine for now.

File: crates/polars-io/src/ipc/object_store_reader.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    // reader as well but I am not sure it is what we want.
+    let get_result = store.get(path).await.map_err(to_compute_err)?;
```
Comment (ritchie46) [inline crates/polars-io/src/ipc/object_store_reader.rs:72]: Here we should split up the work. We should be able download concurrently and the first chunks should be downloaded first.  We should have a `batched` interface that can already get `DataFrame`s of the chunks that are already downloaded.

**Comment by mickvangelderen** [comment]
Closing this in favor of https://github.com/pola-rs/polars/pull/14984

**Pull request closed without merging**